File: 20.Valid-Parentheses.py
# https://leetcode.com/problems/remove-nth-node-from-end-of-list/description/
#
# algorithms
# Easy (33.62%)
# Total Accepted:    437.1K
# Total Submissions: 1.3M
# beats 33.88% of python submissions

from collections import deque


class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = deque()

        for ch in s:
            if ch == '(' or ch == '[' or ch == '{' or len(stack) == 0:
                stack.append(ch)
            elif ch == ')':
                if stack[-1] == '(':
                    stack.pop()
                else:
                    stack.append(ch)
            elif ch == ']':
                if stack[-1] == '[':
                    stack.pop()
                else:
                    stack.append(ch)
            elif ch == '}':
                if stack[-1] == '{':
                    stack.pop()
                else:
                    stack.append(ch)

        return len(stack) == 0

File: 20.Valid-Parentheses.java
// https://leetcode.com/problems/valid-parentheses/
//
// algorithms
// Easy (36.54%)
// Total Accepted: 596,053
// Total Submissions: 1,631,303
// beats 100.0% of java submissions


class Solution {
    public boolean isValid(String s) {
        if (s.equals("")) {
            return true;
        }
        
        LinkedList<Character> stack = new LinkedList<>();
        HashMap<Character, Character> map = new HashMap<>();
        map.put(')', '(');
        map.put(']', '[');
        map.put('}', '{');
        
        for (char ch : s.toCharArray()) {
            if (ch == '(' || ch == '[' || ch == '{') {
                stack.offerFirst(ch);
            } else if (stack.isEmpty() || stack.peekFirst() != map.get(ch)) {
                return false;
            } else {
                stack.pollFirst();
            }
        }
        
        return stack.isEmpty();
    }
}